### PR TITLE
TextScannerImpl with Junit

### PR DESCRIPTION
**Pull request by DianaSlepikaite**
Signed-off-by: Diana Slepikaite <d.slepikaite@gmail.com>

### CODE REVIEW

**Comment by DianaSlepikaite** [comment]
rerun tests

File: galasa-managers-parent/galasa-managers-core-parent/dev.galasa.textscan.manager/src/main/java/dev/galasa/textscan/spi/TextScannerImpl.java [hunk 1]
```diff
@@ -16,88 +23,255 @@
...
+		boolean find = false;
+		int lines = 0;
+		int occurance = 0;
+		if(count<1) {
+			throw new InvalidParameterException("Invalid occurancies number");
+		}
+		while((line =reader.readLine())!=null) {
+			occurance = count;
+			queue.add(line);
+			if(queue.isAtFullCapacity()) {
```
Comment (Michael-Baylis) [inline galasa-managers-parent/galasa-managers-core-parent/dev.galasa.textscan.manager/src/main/java/dev/galasa/textscan/spi/TextScannerImpl.java:108]: simplify the code,  you don't have to wait for the queue to fill before you start checking

File: galasa-managers-parent/galasa-managers-core-parent/dev.galasa.textscan.manager/src/main/java/dev/galasa/textscan/spi/TextScannerImpl.java [hunk 2]
```diff
@@ -16,88 +23,255 @@
...
+				toScan ="";
```
Comment (Michael-Baylis) [inline galasa-managers-parent/galasa-managers-core-parent/dev.galasa.textscan.manager/src/main/java/dev/galasa/textscan/spi/TextScannerImpl.java:109]: Use StringBuffer to build the string to scan

File: galasa-managers-parent/galasa-managers-core-parent/dev.galasa.textscan.manager/src/main/java/dev/galasa/textscan/spi/TextScannerImpl.java [hunk 3]
```diff
@@ -16,88 +23,255 @@
...
+				for(int i=0;i<queue.size();i++) {
+					toScan+=queue.get(i);
+				}
+				if (failPattern != null) {
+					if (failPattern.matcher(toScan).find()) {
+						throw new FailTextFoundException("Fail Pattern '"+failPattern+"' was found");
+					}
+				}
+				Matcher m = searchPattern.matcher(toScan);
```
Comment (Michael-Baylis) [inline galasa-managers-parent/galasa-managers-core-parent/dev.galasa.textscan.manager/src/main/java/dev/galasa/textscan/spi/TextScannerImpl.java:118]: Have to be extremely careful here.   as the found text moves up the fifo queue,  you could be double accounting the found text and finding more occurenances than there actually is.

File: galasa-managers-parent/galasa-managers-core-parent/dev.galasa.textscan.manager/src/main/java/dev/galasa/textscan/spi/TextScannerImpl.java [hunk 4]
```diff
@@ -16,88 +23,255 @@
...
+				queue.poll();
+			}else {
+				toScan ="";
+				for(int i=0;i<queue.size();i++) {
+					toScan+=queue.get(i);
+				}
+			}
+			lines ++;
+		}
+		if (lines<10) {
```
Comment (Michael-Baylis) [inline galasa-managers-parent/galasa-managers-core-parent/dev.galasa.textscan.manager/src/main/java/dev/galasa/textscan/spi/TextScannerImpl.java:136]: dont need this if you start the above from input line 1

File: galasa-managers-parent/galasa-managers-core-parent/dev.galasa.textscan.manager/src/main/java/dev/galasa/textscan/spi/TextScannerImpl.java [hunk 5]
```diff
@@ -16,88 +23,255 @@
...
+		String toScan = "";
+		String string = "";
+		int found =0;
+		boolean find = false;
+		int lines = 0;
+		int count = 0;
+		if(occurrence<1) {
+			throw new InvalidParameterException("Invalid occurancies number");
+		}
+		while((line =reader.readLine())!=null) {
```
Comment (Michael-Baylis) [inline galasa-managers-parent/galasa-managers-core-parent/dev.galasa.textscan.manager/src/main/java/dev/galasa/textscan/spi/TextScannerImpl.java:230]: same issues here as in the scan method.   you should be able to use the same code for both.

File: galasa-managers-parent/galasa-managers-core-parent/dev.galasa.textscan.manager/src/test/resources/loremIpsum.txt [hunk 6]
```diff
@@ -0,0 +1,98 @@
+Lorem ipsum dolor sit amet, consectetur adipiscing elit.
```
Comment (Michael-Baylis) [inline galasa-managers-parent/galasa-managers-core-parent/dev.galasa.textscan.manager/src/test/resources/loremIpsum.txt:1]: Ok,  this isn't working,   I can't validate the junit tests.   Would you please simplify the input text with stuff like
Line 1
Line 2
Line 3 Match 1
etc

**Review by Michael-Baylis — CHANGES_REQUESTED** [review]
Sorry,  need some work,  especially the possibility of double accounting on the inputstream
I will knock up a text file and call to illustrate

**Comment by DianaSlepikaite** [comment]
rerun tests

File: galasa-managers-parent/galasa-managers-core-parent/dev.galasa.textscan.manager/src/main/java/dev/galasa/textscan/spi/TextScannerImpl.java [hunk 7]
```diff
@@ -5,99 +5,385 @@
...
+					while (occurance > 0) {
+						stp=0;
+						find = m.find();
+						if(find) {
+							stp = m.start()+offset;
+							if(!array.contains(stp)) {
+							found++;
+							array.add(stp);
+
+							}
```
Comment (Michael-Baylis) [inline galasa-managers-parent/galasa-managers-core-parent/dev.galasa.textscan.manager/src/main/java/dev/galasa/textscan/spi/TextScannerImpl.java:181]: should check at this point if count is matched and return.

File: galasa-managers-parent/galasa-managers-core-parent/dev.galasa.textscan.manager/src/main/java/dev/galasa/textscan/spi/TextScannerImpl.java [hunk 8]
```diff
@@ -5,99 +5,385 @@
...
+		} catch (IOException e) {
+			throw new TextScanManagerException();
+		}
+		if (lines<10) {
+			toScan ="";
+			for(int i=0;i<queue.size();i++) {
+				toScan+=queue.get(i)+"\n";
+			}
+			scan(toScan,searchPattern,failPattern,count);
+		}else if(found>0&&found<count) {
```
Comment (Michael-Baylis) [inline galasa-managers-parent/galasa-managers-core-parent/dev.galasa.textscan.manager/src/main/java/dev/galasa/textscan/spi/TextScannerImpl.java:200]: dont need any of this if checked earlier

File: galasa-managers-parent/galasa-managers-core-parent/dev.galasa.textscan.manager/src/main/java/dev/galasa/textscan/spi/TextScannerImpl.java [hunk 9]
```diff
@@ -5,99 +5,385 @@
...
+			throw new IncorrectOccurancesException("Was looking for "+count+" instances of '"+searchPattern+"' but found "+ found);
+		}else if(found==0) {
+			throw new MissingTextException("Search Pattern '"+searchPattern+"' was not found");
```
Comment (Michael-Baylis) [inline galasa-managers-parent/galasa-managers-core-parent/dev.galasa.textscan.manager/src/main/java/dev/galasa/textscan/spi/TextScannerImpl.java:203]: this is all you need at the end of the loop

**Review by Michael-Baylis — CHANGES_REQUESTED** [review]
The method needs a big clean up and rework for readability and maintainability reasons

**Comment by DianaSlepikaite** [comment]
rerun tests

**Comment by DianaSlepikaite** [comment]
rerun tests

**Comment by DianaSlepikaite** [comment]
rerun tests

**Review by Michael-Baylis — APPROVED** [review]

**Pull request merged**